xds/internal/balancer/gracefulswitch/gracefulswitchbalancer.go

```diff
@@ -0,0 +1,374 @@
+	// switched from, as the most recent update from the Resolver does not
+	// concern the balancer being switched from.
+	if bal == gsb.balancerCurrent && gsb.balancerPending != nil {
+		return
+	}
+	gsb.cc.ResolveNow(opts)
+}
+
+func (gsb *gracefulSwitchBalancer) removeSubConn(bal balancer.Balancer, sc balancer.SubConn) {
+	gsb.mu.Lock()
+	defer gsb.mu.Unlock()
+	if bal != gsb.balancerCurrent && bal != gsb.balancerPending {
+		return
+	}
+	gsb.cc.RemoveSubConn(sc) // Is this right? Should we really be intercepting this and checking if current or pending? I feel like this will be incorrect on Close() calls
+}
+
+func (gsb *gracefulSwitchBalancer) updateAddresses(bal balancer.Balancer, sc balancer.SubConn, addrs []resolver.Address) {
+	gsb.mu.Lock()
+	// "Call comes from balancer which is neither current or pending"
+	if bal != gsb.balancerCurrent && bal != gsb.balancerPending {
+		gsb.mu.Unlock()
+		return
+	}
+	gsb.mu.Unlock()
+	// close() and clearing from swap() (and thus making this update come from a
+	// balancer neither current or pending) can get interspliced here before
+	// UpdateAddresses() - however, doesn't seem to be a problem with regard to
+	// correctness.
+	gsb.cc.UpdateAddresses(sc, addrs) // Can this call back inline?
+}
+
+func (gsb *gracefulSwitchBalancer) SwitchTo(builder balancer.Builder) error {
+	if gsb.closed.HasFired() {
+		// logger?
+		return errBalancerClosed
+	}
+	gsb.mu.Lock()
+	ccw := &clientConnWrapper{
+		ClientConn: gsb.cc,
+		gsb:        gsb,
+	}
+	newBalancer := builder.Build(ccw, gsb.bOpts)
+	if newBalancer == nil {
+		// Can this even ever happen?
+		return balancer.ErrBadResolverState
+	}
+	ccw.bal = newBalancer
+	if gsb.balancerCurrent == nil {
+		gsb.balancerCurrent = newBalancer
+		gsb.mu.Unlock()
+		return nil
+	}
+	var balToClose balancer.Balancer
+	gsb.pendingState = balancer.State{
+		ConnectivityState: connectivity.Connecting,
+		Picker:            base.NewErrPicker(balancer.ErrNoSubConnAvailable),
+	}
+	// Clean up resources here that are from a previous pending lb.
+	for sc, sb := range gsb.scToSubBalancer {
+		if sb == gsb.balancerPending {
+			gsb.cc.RemoveSubConn(sc)
+			delete(gsb.scToSubBalancer, sc)
+		}
+	}
+	// Need to test this by verifying somehow.
+	balToClose = gsb.balancerPending
+	gsb.balancerPending = newBalancer
+	gsb.mu.Unlock()
+	if balToClose != nil {
+		balToClose.Close()
+	}
+	return nil
+}
+
+// UpdateClientConnState() simply forwards the update to one of it's children.
+func (gsb *gracefulSwitchBalancer) UpdateClientConnState(state balancer.ClientConnState) error {
+	if gsb.closed.HasFired() {
+		// logger?
+		return errBalancerClosed
+	}
+	gsb.mu.Lock()
+	balToUpdate := gsb.balancerPending
+	if balToUpdate == nil {
+		balToUpdate = gsb.balancerCurrent
+	}
+	gsb.mu.Unlock()
+	balToUpdate.UpdateClientConnState(state)
+	return nil
+}
+
+func (gsb *gracefulSwitchBalancer) ResolverError(err error) {
+	if gsb.closed.HasFired() {
+		// Logger?
+		return
+	}
+	gsb.mu.Lock()
```

Comment:
I was expecting this function to look similar to `UpdateClientConnState`?

Reply:
Me and Doug (and I think you) discussed this. The reason we only forward to pending is because if the resolver specified a configuration that specifies a pending balancer, then an error from the resolver doesn't concern the current balancer (similar to why I eat a call from current for resolveNow{}, since the resolver has already "moved on" to the pending balancer, and is logically tied to that balancer.

Reply:
Oh, I didn't mean this function is wrong. What I meant was:
This function has the same logic as `UpdateClientConnState` (the function right above), so I was expecting the code to also look similar (do the same `bal:=pending; if bal==nil { bal=current }; bal.ForwardUpdate()`)

Reply:
Switched.